Add a buffer cache to containers in VirtualizingStackPanel

gentledepp opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
Uses 3 strategies to optimize scrolling performance of a VirtualizedStackPanel

Instead of measuring and arranging every element for every bit of scrolling as depicted here:

![e50e4e59-b5c1-4a46-bf91-d769c159844f](https://github.com/user-attachments/assets/dccd08e8-ea30-4682-bcb8-3e495f2ab293)

We keep additional elements in memory and only measure and arrange when we reach the _extended viewport_
Also, we only measure and arrange elements that have been _changed_!

![582e5fb9-7a37-48fa-9bb5-77f7da7f354c](https://github.com/user-attachments/assets/4905f468-9322-4e90-be2b-f6a0fa6fbbec)

The downside of this approach is, that we use more memory (as we need to keep additional items in memory)
The upside is, that re _greatly_ reduce the number of measure & arrange cycles which reduces the stress on the GC

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
Scrolling on Android is very edgy (non-smooth) since many Measure/Arrange calls put a lot of pressure on the Garbage Collector:

![perf_1](https://github.com/user-attachments/assets/dea3d9ad-1fd3-49df-ad87-b308d0a1fb1c)

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
Smooth scrolling on all platforms, even with more complex/deeper nested list item layouts.

![perf_2](https://github.com/user-attachments/assets/6ade1a3e-9d86-4882-a322-5d4b0f6bff1c)


## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->

The default implementation of the `VirtualizingStackPanel` only keeps enough visual elements in memory to fill its current bounds. This uses the least amount of memory.

On devices such as Android, however, it leads to rough/edgy (non-smooth) scrolling.
This is because for every bit of scrolling, all elements of the `VirtualizingStackPanel` need to be measured and arranged, which causes extensive GC/memory pressure on Android.

The approach implemented here tries to keep the number of Measure/Arrange cycles as low as possible.

This is achieved by applying two strategies:

### 1 - Buffer more items (the first commit of this PR)
In order to not have to recycle elements every time the user scrolls "a bit", we extended the viewport by a `BufferFactor`.
By default this is 0.5 - so if the `VirtualizingStackPanel` scrolls vertically and can show 800 px, we do not only render elements for 800 px, but also 400 (800*BufferFactor) above and below
Only when reaching the edge of these *additional elements*, we detect and call `InvalidateMeasure`.
This causes more memory to be used (since twice as many UI elements are kept in memory), but greatly reduces the number of `Measure` calls.

The second commit of this PR handles a corner case:
When we scroll near the end or start of the list, the `VirtualizingStackPanel` keeps trying to fill the available space.
Since no more element is there to be rendered, however, this space cannot be filled. Not being aware of this causes the panel to call `InvalidateMeasure` on every scroll (towards the start and end of the list)
You can see this in the log output below since the last 4 logs call `InvalidateMeasure` 
```
	old xVP:374,8.-1675,6 - new VP:649,6-1300
	old xVP:374,8.-1675,6 - new VP:600-1250,4
	old xVP:374,8.-1675,6 - new VP:550,4-1200,8
	old xVP:374,8.-1675,6 - new VP:500-1150,4
	old xVP:374,8.-1675,6 - new VP:449,6-1100
	old xVP:374,8.-1675,6 - new VP:400-1050,4
	old xVP:374,8.-1675,6 - new VP:350,4-1000,8 Case 1a => InvalidateMeasure
	old xVP:25,2.-1326 - new VP:300-950,4
	old xVP:25,2.-1326 - new VP:249,6-900
	old xVP:25,2.-1326 - new VP:200-850,4
	old xVP:25,2.-1326 - new VP:150,4-800,8 Case 1b UP => InvalidateMeasure    
	old xVP:0.-1126 - new VP:100-750,4 Case 1b UP => InvalidateMeasure
	old xVP:0.-1075,6 - new VP:49,6-700 Case 1b UP => InvalidateMeasure
	old xVP:0.-1025,2 - new VP:0-650,4 Case 1b UP => InvalidateMeasure
```
After implementing the strategy for this edge case, it looks like this:
```
	old xVP:399,2.-2350,4 - new VP:600-1250,4
	old xVP:399,2.-2350,4 - new VP:550,4-1200,8
	old xVP:399,2.-2350,4 - new VP:500-1150,4
	old xVP:399,2.-2350,4 - new VP:449,6-1100
	old xVP:399,2.-2350,4 - new VP:400-1050,4
	old xVP:399,2.-2350,4 - new VP:350,4-1000,8 Case 1a => InvalidateMeasure
	old xVP:0.-1651,2 - new VP:300-950,4 Case 1b UP
	old xVP:0.-1651,2 - new VP:249,6-900 Case 1b UP
	old xVP:0.-1651,2 - new VP:200-850,4 Case 1b UP
	old xVP:0.-1651,2 - new VP:150,4-800,8 Case 1b UP
	old xVP:0.-1651,2 - new VP:100-750,4 Case 1b UP
	old xVP:0.-1651,2 - new VP:49,6-700 Case 1b UP
	old xVP:0.-1651,2 - new VP:0-650,4 Case 1b UP
```
You can see that `InvalidateMeasure` was called and then the effective viewport changed 7 consecutive times without any `InvalidateMeasure` call. 
This is because the first call detected that it is a t the start of the list and re-measuring would have no effect at all.

### 2 - Avoiding measuring realized elements when only the scroll position changed

The third commit of this PR implements another optimization:
When we _only_ scroll, it makes no sense to call `Measure` on any elements that will be kept in the viewport since they did not change at all.
Therefore we try to detect 
  - that _only_ the scroll position changed and 
  - whether a ui element was kept (not created or re-used)
...and in this case skipt the `Measure` call.

**NOTE** as I am not a 100% "Avalonia native developer", please check this code carefully since I do _not_ know if all edge-cases are handled correctly!!

However, when implemented, the output can look something like:
```
	old xVP:0.-1300,8 - new VP:49,6-700
	old xVP:0.-1300,8 - new VP:100-750,4
	old xVP:0.-1300,8 - new VP:150,4-800,8
	old xVP:0.-1300,8 - new VP:200-850,4
	old xVP:0.-1300,8 - new VP:249,6-900
	old xVP:0.-1300,8 - new VP:300-950,4
	old xVP:0.-1300,8 - new VP:350,4-1000,8
	old xVP:0.-1300,8 - new VP:400-1050,4
	old xVP:0.-1300,8 - new VP:449,6-1100
	old xVP:0.-1300,8 - new VP:500-1150,4
	old xVP:0.-1300,8 - new VP:550,4-1200,8
	old xVP:0.-1300,8 - new VP:600-1250,4
	old xVP:0.-1300,8 - new VP:649,6-1300
	old xVP:0.-1300,8 - new VP:700-1350,4 Case 1a => InvalidateMeasure
	Skipped measuring Item 1
	Skipped measuring Item 2
	Skipped measuring Item 3
	Skipped measuring Item 4
	Skipped measuring Item 5
	Skipped measuring Item 6
	Skipped measuring Item 7
	Skipped measuring Item 8
	Skipped measuring Item 9
	Skipped measuring Item 10
	Skipped measuring Item 11
	Skipped measuring Item 12
	Skipped measuring Item 13
	Skipped measuring Item 14
	Skipped measuring Item 15
	Skipped measuring Item 16
	Skipped measuring Item 17
	Skipped measuring Item 18
	Skipped measuring Item 19
	Skipped measuring Item 20
	Skipped measuring Item 21
	Skipped measuring Item 22
	Skipped measuring Item 23
	Skipped measuring Item 24
	Skipped measuring Item 25
	Skipped measuring Item 26
	Skipped measuring Item 27
	Skipped measuring Item 28
	Skipped measuring Item 29
	Skipped measuring Item 30
	Skipped measuring Item 31
	Skipped measuring Item 32
	old xVP:49,5999999999999.-2000,8 - new VP:750,4-1400,8

```

Finally I thought: Why not also cache the "arranged" elements if we only scroll?
So I added that too. ;-)

**NOTE** I also introduced a readonly variable ` private readonly bool _traceVirtualization`
If you set it to `true` you get the same debug output as above which will help you understand the implementation.


## Checklist

- [ ] Added unit tests (if possible)? - I don't know where and how. Please advise!
- [x] Added XML documentation to any related classes? 
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation - I'd happily do so. But only `BufferFactor` needs to be documented and I'd like to have the implementation and naming accepted first

## Breaking changes
<!--- List any breaking changes here. -->
None - we can set the default `BufferFactor` to `0` and the implementation will essentially be the same as before this PR.

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
Implements #18626

Commit:
#18626 - improved scrolling performance in VirtualizingStackPanel.cs by reducing Measure/Arrange calls since they cause heavy GC pressure on constrained devices (Android, iOS) especially with complex item views

MrJul commented:
First of all, thank you for your contribution!

I haven't looked at the code at all yet, but there's a bunch of failing tests after this PR: https://dev.azure.com/AvaloniaUI/AvaloniaUI/_build/results?buildId=56039&view=ms.vss-test-web.build-test-results-tab. Some existing tests might need adjustments while others look like real failures that should be addressed.

Moreover, the new feature should come with its own unit tests validating the new expected behaviors regarding realization, scrolling and measuring. `VirtualizingStackPanel` has complex logic, and we shouldn't add anything new here that would be easily broken in the future without tests.

cla-avalonia commented:
@alexander.marek,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

**alexander.marek** doesn't seem to be a GitHub user.
Please add the emails used in your commits to your Github account accordingly.

gentledepp commented:
@MrJul  you are right.
I fixed most of the the tests yesterday night. 
They failed because the default `BufferFactor` was set to 0.5, which obviously changed the panels behavior.

Could you please assist me in fixing one more test?
`Avalonia.Controls.UnitTests.VirtualizingStackPanelTests.Recycling_A_Hidden_Control_Shows_It`
```csharp
        [Fact]
        public void Recycling_A_Hidden_Control_Shows_It()
        {
            using var app = App();
            var style = CreateIsVisibleBindingStyle();
            var items = Enumerable.Range(0, 3).Select(x => new ItemWithIsVisible(x)).ToList();
            var (target, scroll, itemsControl) = CreateTarget(items: items, styles: new[] { style });
            var container = target.ContainerFromIndex(2)!;

            Assert.True(container.IsVisible);
            Assert.Equal(20, container.Bounds.Top);

            items[2].IsVisible = false;
            Layout(target);

            Assert.False(container.IsVisible);

            items.RemoveAt(2);
            items.Add(new ItemWithIsVisible(3));
            Layout(target); //THIS DOES NOT CALL MEASUREOVERRIDE ANYMORE! 

            Assert.True(container.IsVisible);
        }
```
For some reason, the `MeasureOverride` call of the `VirtualizingStackPanel` is not called anymore after applying my first commit.
I honestly do not understand why that is the case.

But I also do not understand the test.
If an item changes, it should re-render the container. 
OK.
But the ItemsSource is a simple List<T> - so withotu any NotifyingCollectionChanged.

Would it be feasible if I changed this to be an `ObservableCollection`


I will add the tests for the buffer today then!

gentledepp commented:
@MrJul please wait with review - I am still adding specialized tests for the bufferfactor!

gentledepp commented:
@cla-avalonia agree

gentledepp commented:
@MrJul some test are failing, but it seems to be an appium issue (not related to my changes - i guess?)

> OpenQA.Selenium.WebDriverException : An unknown server-side error occurred while processing the command. Original error: 'POST /element' cannot be proxied to Mac2 Driver server because its process is not running (probably crashed)

Please advise

gentledepp commented:
Are there any plans for reviewing this?

Alternatively, is there any guide on how to create your own nuget package of avalonia locally?
I do not want to spam you with pseudo PRs, just to get a usable package :-D

grokys commented:
Hi @gentledepp - I definitely plan to review this, the only problem is a lack of time - sorry. Thank you for the great PR writeup, it should make reviewing it much easier!

grokys commented:
Hi @gentledepp - I hope to review this today.

> Alternatively, is there any guide on how to create your own nuget package of avalonia locally? I do not want to spam you with pseudo PRs, just to get a usable package :-D

Yes, you can create local nuget packages quite easily using the `BuildToNuGetCache` nuke target which will push a set of `9999.0.0-localbuild` nuget packages to your nuget cache: https://github.com/AvaloniaUI/Avalonia/pull/14446

grokys commented on src/Avalonia.Controls/VirtualizingStackPanel.cs:
```diff
@@ -185,13 +238,15 @@ protected override Size MeasureOverride(Size availableSize)
             }
             finally
             {
+                _cacheElementMeasurements = false;
```
Took me a little which to understand why `_cacheElementMeasurements` is reset here, but if I remove this line, then `Focused_Container_Is_Positioned_Correctly_when_Container_Size_Change_Causes_It_To_Be_Moved_Into_Visible_Viewport` fails, so I *think* that the reason for it is that it's only set to true after a viewport change due to scrolling, and if a measure is invoked for any other reason then it needs to be false? If so might be worth adding a quick comment here to explain it (as the rest of the code is so well commented).

gentledepp replied:
yeah that is exactly the idea.
My problem was, that I wanted to make scrolling fast in a virtualizingstackpanel (as fast as it is in a normal stackpanel) so I tried caching as much as possible.

Doesn't a call to "InvalidateMeasure" invalidate all measurements down the visual tree?  At least I thought it did.
One reason why the normal "VirtualizingStackPanel" is so slow (for complex item layouts) is that it measures everything anytime a new item is shown.
So I _assume_ the caching you are talking about is at least not working? (if it _is_, I am happy to remove that code and the complexity that comes with it)

grokys commented on src/Avalonia.Controls/VirtualizingStackPanel.cs:
```diff
@@ -217,8 +272,22 @@ protected override Size ArrangeOverride(Size finalSize)
+
+                        // Always arrange the element, even if it's outside the visible viewport
+                        // but not if it is an unchanged element
+                        if(!_unchangedElements.Contains(e))
```
The code goes to a lot of effort to ensure that measure isn't called unnecessarily on containers, but I wonder if some of the mechanisms used such as this `_unchangedElements` `HashSet` are worse for performance than just calling `Measure`? 

Measuing should be a no-op if the [constraints haven't changed](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Layout/Layoutable.cs#L371)

gentledepp replied:
please ask your specialists if I am indeed wrong here.
I think: If the root view calls InvalidateMeasure, all child-measures are also invalidated.
If I am wrong, I am going to remove this caching mechanism!

grokys replied:
Hey @gentledepp - nope, `InvalidateMeasure` only [invalidates the control itself](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Layout/Layoutable.cs#L446-L451) and not descendants. It does however queue a layout pass and which has overhead and should be avoided if possible.

However, this `_unchangedElements` isn't preventing an `InvalidateMeasure` from what I can see? [It's preventing an `Arrange` call](https://github.com/AvaloniaUI/Avalonia/pull/18646/files/5caef7aee48c7e1d333cefd410e880ea49cf13b4#diff-5d79415cbb9133eebf6f06a48b4fb75aff8f14854e4b66db7dc49915d52221d1R279), which has very little overhead if the constraints have not changed. It could be that I'm missing where the `InvalidateMeasure` gets called though.

(Apologies: I said it was preventing a `Measure` call before, but it's actually `Arrange`. Both should be no-ops if the constraints haven't changed and the measure/arrange is still valid though).

gentledepp replied:
Ok. I trust your expertise in that.
I removed the measure/arrange caching and the associated tests

grokys replied:
One way to unit test this would be to use a custom container type: `MeasureOverride`/`ArrangeOverride` shouldn't get called if the measure/arrange was a no-op.

gentledepp replied:
you mean ItemsContainer?

gentledepp replied:
isn't that exactly what I did?
```



        private class VirtualizingStackPanelCountingMeasureArrange : VirtualizingStackPanel
        {
            public int Measured { get; set; }
            public int Arranged { get; set; }

            public void ResetMeasureArrangeCounters()
            {
                // reset counters
                Measured = 0;
                Arranged = 0;
                foreach (var itm in Items.OfType<ICountMeasureArrangeCalls>())
                {
                    itm.Measured = 0;
                    itm.Arranged = 0;
                }
            }

            protected override Size MeasureOverride(Size availableSize)
            {
                Measured++;
                return base.MeasureOverride(availableSize);
            }

            protected override Size ArrangeOverride(Size finalSize)
            {
                Arranged++;
                return base.ArrangeOverride(finalSize);
            }

            protected internal override void MeasureElement(Control element, Size availableSize)
            {
                if (element is ContentPresenter cp && cp.Content is ICountMeasureArrangeCalls itm)
                    itm.Measured++;

                base.MeasureElement(element, availableSize);
            }

            protected internal override void ArrangeElement(Control element, Rect rect)
            {
                if (element is ContentPresenter cp && cp.Content is ICountMeasureArrangeCalls itm)
                    itm.Arranged++;

                base.ArrangeElement(element, rect);
            }
        }
        
        ```
        
Or do you mean: Derive from ContentPresenter (`MACountingPresenter: ContentPresenter`) , then override Measure and arrange and count it?

gentledepp replied:
I guess I know what you mean now:

I am using specialized callbacks ` protected internal override void MeasureElement(Control element, Size availableSize)` just to test this.
However, this only shows us how often the stackpanel calls measure on the elements.

But if the elements themselves cache their measure/arranges, the tests do not cover that.

So i changed the tests to use a `TestCanvas` that counts the real measure/arrange calls:
```
private static FuncDataTemplate<ItemWithHeight> CanvasWithHeightTemplate = new((_, _) =>
    new TestCanvas
    {
        Width = 100,
        [!Layoutable.HeightProperty] = new Binding("Height"),
    });

private static FuncDataTemplate<ItemWithWidth> CanvasWithWidthTemplate = new((_, _) =>
    new TestCanvas
    {
        Height = 100,
        [!Layoutable.WidthProperty] = new Binding("Width"),
    });

public class TestCanvas : Canvas
{
    protected override Size MeasureOverride(Size availableSize)
    {
        if(DataContext is ICountMeasureArrangeCalls itm)
            itm.Measured++;

        return base.MeasureOverride(availableSize);
    }

    protected override Size ArrangeOverride(Size finalSize)
    {
        if (DataContext is ICountMeasureArrangeCalls itm)
            itm.Arranged++;

        return base.ArrangeOverride(finalSize);
    }
}
```

And removed my caching implementation.
Result: The tests still succeeded. So you were right and I can remove that caching code altogether!

grokys commented on src/Avalonia.Controls/VirtualizingStackPanel.cs:
```diff
@@ -74,11 +93,35 @@ public class VirtualizingStackPanel : VirtualizingPanel, IScrollSnapPointsInfo
+        private bool _cacheElementMeasurements = false;
+        private Size _previousSize;
+        private readonly HashSet<Control> _unchangedElements = new();
+        private readonly bool _traceVirtualization = false;
```
Before we merge this PR, we should remove the debug code - I can understand that it's useful for debugging, but we don't have debug code lying around anywhere else. Hopefully it should all be covered by unit tests by now.

gentledepp replied:
absolutely - as mentioned in my PR, i just kept this for you to follow along more easily

grokys commented on src/Avalonia.Controls/VirtualizingStackPanel.cs:
```diff
@@ -51,7 +52,25 @@ public class VirtualizingStackPanel : VirtualizingPanel, IScrollSnapPointsInfo
+                validate: v => v is >= 0 and <= 2);
 
+        /// <summary>
+        /// Gets or sets the factor to determine how much additional space to maintain above and below the viewport.
```
Nit: `<summary>` sections should be a single sentence with the remainder going into `<remarks>` (i can't find the docs for this right now, but if you look at the MS API docs you'll see that is generally the guideline)

grokys commented on src/Avalonia.Controls/VirtualizingStackPanel.cs:
```diff
@@ -51,7 +52,25 @@ public class VirtualizingStackPanel : VirtualizingPanel, IScrollSnapPointsInfo
+        /// This uses more memory as more UI elements are realized, but greatly reduces the number of Measure-Arrange
+        /// cycles which can cause heavy GC pressure depending on the complexity of the item layouts.
+        /// </summary>
+        public double BufferFactor
```
Nit: Put properties with other properties (not at the top of the file with static fields)

grokys commented on tests/Avalonia.Controls.UnitTests/VirtualizingStackPanelTests.Buffered.cs:
```diff
@@ -0,0 +1,2321 @@
+            var items = Enumerable.Range(0, 100).Select(x => new ItemWithHeight(x)).ToList();
+            
+            // Act
+            var (target, scroll, itemsControl) = CreateTarget<ItemsControl,VirtualizingStackPanelCountingMeasureArrange>(items: items, itemTemplate: CanvasWithHeightTemplate);
```
Nit: Some lines are extremely long (>210 chars): please follow the [style guidelines](https://github.com/AvaloniaUI/Avalonia/blob/master/CONTRIBUTING.md#style) (we don't have many). I have a 4k monitor and I need to scroll to see the end of these lines ;)

gentledepp replied:
I introduced line breaks

grokys commented on src/Avalonia.Controls/VirtualizingStackPanel.cs:
```diff
@@ -51,7 +52,25 @@ public class VirtualizingStackPanel : VirtualizingPanel, IScrollSnapPointsInfo
+        /// <summary>
+        /// Defines the <see cref="BufferFactor"/> property.
+        /// </summary>
+        public static readonly StyledProperty<double> BufferFactorProperty =
```
One thing to consider: WinUI/UWP has a similar concept as this in `ItemsRepeater` and the properties which control this behavior there are called `HorizontalCacheLength` and [`VerticalCacheLength`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.itemsrepeater.verticalcachelength?view=windows-app-sdk-1.7).

As `VirtualizingStackPanel` can only virtualize on one axis, we don't need two properties but we should consider whether we want to adopt similar naming? We'll do an API review on this at the end anyway.

gentledepp replied:
I do not have a preference on the naming.
I thing you guys are the owners and should therefore tell me how to name stuff.

That being said: I wanted to use a relative value, therefore called "factor"
`Length` sounds like a fixed pixel length that is used for caching. 

So, we could call it `VirtualizationFactor` or `CachingFactor`?

grokys replied:
Not sure just yet - we probably need to do a quick API review with the core team.

gentledepp replied:
please do :-)

grokys reviewed (commented):
Hi @gentledepp!

First of all sorry that it took me so long to review, you caught me at a time where I was busy with some other stuff and I'm the only one who really knows this part of the codebase well enough to give a review.

Secondly, thank you so much for the fantastic PR description - it's one of the best I've ever seen and was a joy to read ;)

Thirdy, thank you for the code - it's a really useful feature. I've tried to dig into the code as best I can and I think I understand what it's doing reasonably well now. The only real concerns I have about this PR relate to the duplication of tests and whether `_unchangedElements` is really needed. The rest of the comments are more nits, and I ask forgiveness for these; I'd be quite happy to tidy these up myself before merge (let me know) ;)

gentledepp commented on tests/Avalonia.Controls.UnitTests/VirtualizingStackPanelTests.Buffered.cs:
```diff
@@ -0,0 +1,2321 @@
+        }
+        
+        [Fact]
+        public void Creates_Initial_Items()
```
sure. I will remove all duplicate tests.
Some, though, should behave differently because of the `BufferFactor` and will be kept

gentledepp replied:
so.. i merged and parameterized them. I hope it is fine now?

gentledepp commented on src/Avalonia.Controls/VirtualizingStackPanel.cs:
```diff
@@ -74,11 +93,35 @@ public class VirtualizingStackPanel : VirtualizingPanel, IScrollSnapPointsInfo
             _updateElementIndex = UpdateElementIndex;
+
+            _bufferFactor = Math.Max(0, BufferFactor);
+            // Initialize extended viewport with empty rect
```
yeah.. classic "captain obvious move"... my apologies

gentledepp commented:
@grokys I incorporated your feedback.
What is missing:
1. deciding on the naming of the "BufferFactor" property
2. if caching the measure-arrange calls really makes sense. Otherwise I will remove these calls

maxkatz6 commented:
We are going to have next api review early next month most at earliest.

grokys commented:
@gentledepp rather than make you wait until the next API review session, we're doing an out of band review for this one property right now.

grokys commented:
Hi @gentledepp - we did a quick mini API  review and the best name we could collectively think of was `CacheLength`:

- Mainly because of the precedent of `HorizontalCacheLength` and `VerticalCacheLength` in WinUI/UWP
- It's true that "factor" is a nice descriptor, but we prioritize consistency with other XAML frameworks where their naming isn't _completely_ off

What do you think? If you'd like a second opinion we can wait until @MrJul gets back from his vacation and do one of our proper reviews.

gentledepp commented:
I totally disagree.
Basically, keeping names consistent makes total sense.
However, in WinUI/UWP those Lengths are defined as:
> Gets or sets a value that indicates the size of the buffer used to realize items when panning or scrolling vertically.
So this is a size in _pixels_ I assume.

My implementation, on the other hand, uses a _factor_ to relatively calculate the area of additionally cached items.
So `CacheFactor` seems to be a more suitable name.

Why would you even keep a consistent naming? I guess: To make it easier to developers coming from WinUI and UWP to get used to Avalonia.
Now, using _CacheLength_ will cause the opposite: Developers will enter a fixed number (e.g. 100) and be "surprised" that it behaves differently.

Should I be, however, wrong and the `VerticalCacheLength` and `HorizontalCacheLength` properties are also factors in UWP and WinUI, your name is absolutely the best choice.

That being said: It is your framework - so you decide. ;-)

gentledepp commented:
@grokys has been a while already. Any plans to merge this PR? 😅

danwalmsley commented:
@gentledepp I noticed the code dealing with “UnchangedElements” to skip measuring for items that were previously visible was removed, was this not needed in the end?

danwalmsley commented:
@gentledepp is my understanding correct that the performance improvement here is because previously it would do recycling as each item moved passed the viewport, but now it doesnt even have to do any recycling until it scrolls 25% of the items (if 25% were before and 25% after) meaning its doing the recycling in say chunks of say 10 items at once. meaning that you have to run that code less often. (Essentially it ends up doing the recycling in small could you call them pages?)

danwalmsley reviewed (changes requested):
found an unused field?

gentledepp commented:
> @gentledepp is my understanding correct that the performance improvement here is because previously it would do recycling as each item moved passed the viewport, but now it doesnt even have to do any recycling until it scrolls 25% of the items (if 25% were before and 25% after) meaning its doing the recycling in say chunks of say 10 items at once. meaning that you have to run that code less often. (Essentially it ends up doing the recycling in small could you call them pages?)
> 
> I understand the measure is cached… but why cant the non-visible element arrange be skipped?
> 
> ```
> // Always arrange the element, even if it's outside the visible viewport
>                          ArrangeElement(e, rect);
> ```
> 
> BTW im asking these questions with a view to port your optimizations over to TreeDataGrid, its code is more or less based on this same code too.

yes it is.
Especially on Android I noted that every measure/arrange causes a lot of GC activity which in the end kills the scrolling performance. So I tried to reduce these calls as much as possible

TomEdwardsEnscape commented:
I have two points I'd like to raise.

## Caching strategies

I implemented caching in a panel I created for my product. But it doesn't work like this:

> ![582e5fb9-7a37-48fa-9bb5-77f7da7f354c](https://private-user-images.githubusercontent.com/1175751/434346904-4905f468-9322-4e90-be2b-f6a0fa6fbbec.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAyNjIyMTEsIm5iZiI6MTc1MDI2MTkxMSwicGF0aCI6Ii8xMTc1NzUxLzQzNDM0NjkwNC00OTA1ZjQ2OC05MzIyLTRlOTAtYmUyYi1mNmEwZmE2ZmJiZWMuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MThUMTU1MTUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTQ3M2U2MzBlYTAyZjk1YjA2ZWMwYjdlZDM2MWQ1MTU0ODVhMGMyZWMwYTMwNDNlZmExYjU1ZTZlYWNmMjI4MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.YPBQ0o_oq-TKFYctBS-hz-aAkhi7qYIBqstdbtJKIcA) 

Instead it continues to realise controls one-by-one. It just does so outside the viewport area. This is because each time an item is realised in this panel we start asynchronously loading data from an external source (either a web service or a local cache), and this always takes at least one frame to complete, often many more. With the caching strategy I implemented we can generally avoid the user seeing items before they have finished their async loading process.

The strategy demonstrated in the image above is is of little use for the problem my product needs to solve because the cache "buffer" frequently runs out. Then the user starts seeing items before they have fully loaded. You can reduce how frequently the buffer runs out by increasing the cache length, but this is costly as more and more data need to be loaded as the cache lengthens, and it will still hit zero every so often.

Is it not enough to avoid re-measuring controls that weren't created/recycled? Do we really reduce overhead within each individual measure call by executing more of them in the same layout pass?

## Cache Length Unit

[WPF lets us configure the unit of cache length.](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.virtualizationcachelengthunit?view=windowsdesktop-9.0) You can use `Item`, set an integer cache length, and know that this means that exactly N items will be cached. This is very useful when using logical scrolling. Can support for this be added?

grokys commented:
@TomEdwardsEnscape re: your comments:

## Caching strategies

I think you meant to share a screenshot but nothing is showing up. Anyway I think what you're describing addresses a different use-case. This PR is meant to solve the common problem of scrolling stuttering in a fully-loaded list of items. 

> Is it not enough to avoid re-measuring controls that weren't created/recycled?

I believe the problem is not re-measuring of controls that weren't created/recycled, and instead the fact that every e.g. 25 pixels, a new container is realised, which triggers a layout pass. From a previous comment from the author: "Especially on Android I noted that every measure/arrange causes a lot of GC activity which in the end kills the scrolling performance."

## Cache Length Unit

What do you mean by logical scrolling here? This property is implemented on `VirtualizingStackPanel` which handles its own virtualization and doesn't implement `ILogicalScrollable`.

TomEdwardsEnscape commented:
> @TomEdwardsEnscape this sound a bit confusing to me. Why not do the obvious:
> 
> * for create viewmodels, and asynchronously ensure their data is loaded compeltely
> * thereafter add it to the "ItemsSource"
> 
> am I missing something? 😅

This defeats the purpose of virtualisation. We don't want to download 50 query response pages and 5000 thumbnails from the internet before we display anything to the user, any more than we want to immediately generate 5000 containers.

It seems from your other comments that "not measuring" the other containers isn't actually doing anything, so the performance gains do indeed come entirely from batching up container generation. I don't see a way to unify behaviour to satisfy both sets of requirements given this. We'll just have to use different panels for the different scenarios.

> What do you mean by logical scrolling here? This feature is implemented on `VirtualizingStackPanel` which handles its own virtualization and doesn't implement `ILogicalScrollable`.

Logical scrolling is just an example of a scenario in which caching by item becomes more important. But it's always useful IMO.

grokys committed:
Merge branch 'master' into feature/18626_virtualizingstackpanelperf2

grokys committed:
Merge branch 'master' into feature/18626_virtualizingstackpanelperf2

MrJul committed:
Merge branch 'master' into feature/18626_virtualizingstackpanelperf2
